File: tests/python/extensions/pybind/geometry/rdr2geo.py
#!/usr/bin/env python3

import itertools
import os
import types

from osgeo import gdal
import numpy as np
import pytest

import iscetest
import isce3
from isce3.geometry import (compute_east_north_ground_to_sat_vector,
                            Rdr2GeoParams)
from nisar.products.readers import SLC


def test_point():
    # Subset of tests/cxx/isce3/geometry/geometry/geometry.cpp
    fn = os.path.join(iscetest.data, "envisat.h5")
    slc = SLC(hdf5file=fn)
    orbit = slc.getOrbit()
    subband = "A"
    doplut = slc.getDopplerCentroid(frequency=subband)
    grid = slc.getRadarGrid(frequency=subband)

    # First row of input_data.txt
    dt = isce3.core.DateTime("2003-02-26T17:55:22.976222")
    r = 826988.6900674499
    h = 1777.0

    dem = isce3.geometry.DEMInterpolator(h)
    t = (dt - orbit.reference_epoch).total_seconds()
    dop = doplut.eval(t, r)
    wvl = grid.wavelength

    # native doppler, expect first row of output_data.txt
    llh = isce3.geometry.rdr2geo(t, r, orbit, grid.lookside, dop, wvl, dem)
    assert np.isclose(np.degrees(llh[0]), -115.44101120961082)
    assert np.isclose(np.degrees(llh[1]), 35.28794014757191)
    assert np.isclose(llh[2], 1777.0)

    # zero doppler, expect first row of output_data_zerodop.txt
    llh = isce3.geometry.rdr2geo(t, r, orbit, grid.lookside, 0.0, dem=dem)
    assert np.isclose(np.degrees(llh[0]), -115.43883834023249)
    assert np.isclose(np.degrees(llh[1]), 35.29610867314526)
    assert np.isclose(llh[2], 1776.9999999993)

    # try with bracketing method
    xyz = isce3.geometry.rdr2geo_bracket(t, r, orbit, grid.lookside, 0.0, 0.0,
                                         dem=dem)
    llh = isce3.core.Ellipsoid().xyz_to_lon_lat(xyz)
    assert np.isclose(np.degrees(llh[0]), -115.43883834023249)
    assert np.isclose(np.degrees(llh[1]), 35.29610867314526)
    assert np.isclose(llh[2], 1776.9999999993)

    # Run again without dem argument.  This is just to catch a former bug where
    # EPSG code was uninitialized.  Can't check answer since height is different
    # from default ctor.
    xyz = isce3.geometry.rdr2geo_bracket(t, r, orbit, grid.lookside, 0.0, 0.0)

    # Verify look side parsing
    with pytest.raises(ValueError):
        isce3.geometry.rdr2geo_bracket(t, r, orbit, "Foot", 0.0, 0.0)


@pytest.fixture(scope="module")
def unit_test_params():
    params = types.SimpleNamespace()

    # prepare Rdr2Geo init params
    h5_path = os.path.join(iscetest.data, "envisat.h5")

    radargrid = isce3.product.RadarGridParameters(h5_path)
    params.radargrid = radargrid

    params.slc = SLC(hdf5file=h5_path)
    orbit = params.slc.getOrbit()
    doppler = params.slc.getDopplerCentroid()

    ellipsoid = isce3.core.Ellipsoid()

    # init Rdr2Geo class
    params.rdr2geo_obj = isce3.geometry.Rdr2Geo(
        params.radargrid, orbit, ellipsoid, doppler
    )

    # load test DEM
    params.dem_raster = isce3.io.Raster(
        os.path.join(iscetest.data, "srtm_cropped.tif")
    )
    dem_interp = isce3.geometry.DEMInterpolator(params.dem_raster)
    dem_interp.compute_min_max_mean_height()

    # golden data layer tolerances
    params.tols = [1.0e-5, 1.0e-5, 0.15, 1.0e-4, 1.0e-4, 0.02, 0.02]

    # tolerances for east/north ground to satellite ENU unit vector layers
    # use same tolerance as heading angle since the computations are similar
    params.sat_to_ground_tols = [1.0e-4, 1.0e-4]

    return params


def test_run(unit_test_params):
    """
    check if topo runs with automatically generated layers
    """
    # run
    unit_test_params.rdr2geo_obj.topo(unit_test_params.dem_raster, ".")


def test_run_raster_layers(unit_test_params):
    """
    check if topo runs with manually set layers
    """
    radargrid = unit_test_params.radargrid
    length, width = radargrid.shape

    # list of tuples of file name w/o extension and gdal type
    fnames_dtypes = [
        ("x", gdal.GDT_Float64),
        ("y", gdal.GDT_Float64),
        ("z", gdal.GDT_Float64),
        ("inc", gdal.GDT_Float32),
        ("hdg", gdal.GDT_Float32),
        ("localInc", gdal.GDT_Float32),
        ("localPsi", gdal.GDT_Float32),
        ("simamp", gdal.GDT_Float32),
        ("layoverShadowMask", gdal.GDT_Byte),
        ("groundToSatEast", gdal.GDT_Float32),
        ("groundToSatNorth", gdal.GDT_Float32),
    ]

    # prepare test rasters
    [
        x_raster,
        y_raster,
        height_raster,
        incidence_angle_raster,
        heading_angle_raster,
        local_incidence_angle_raster,
        local_psi_raster,
        simulated_amplitude_raster,
        layover_shadow_raster,
        ground_to_sat_east_raster,
        ground_to_sat_north_raster,
    ] = [
        isce3.io.Raster(f"{fname}.rdr", width, length, 1, dtype, "ENVI")
        for fname, dtype in fnames_dtypes
    ]

    # run with prepared test rasters
    unit_test_params.rdr2geo_obj.topo(
        unit_test_params.dem_raster,
        x_raster,
        y_raster,
        height_raster,
        incidence_angle_raster,
        heading_angle_raster,
        local_incidence_angle_raster,
        local_psi_raster,
        simulated_amplitude_raster,
        layover_shadow_raster,
        ground_to_sat_east_raster,
        ground_to_sat_north_raster,
    )

    # combine all test rasters in a VRT
    _ = isce3.io.Raster(
        "topo_layers.vrt",
        raster_list=[
            x_raster,
            y_raster,
            height_raster,
            incidence_angle_raster,
            heading_angle_raster,
            local_incidence_angle_raster,
            local_psi_raster,
            simulated_amplitude_raster,
            ground_to_sat_east_raster,
            ground_to_sat_north_raster,
        ],
    )


def test_validate(unit_test_params):
    """
    validate generated results
    """
    # load reference topo raster
    ref_ds = gdal.Open(
        os.path.join(iscetest.data, "topo/topo.vrt"), gdal.GA_ReadOnly
    )
    golden_bands = range(1, ref_ds.RasterCount + 1)
    east_north_vec_bands = [8, 9]

    for test_path in ["topo.vrt", "topo_layers.vrt"]:
        # load generated topo raster
        test_ds = gdal.Open(test_path, gdal.GA_ReadOnly)

        # loop thru bands found in golden data and check tolerances
        for i_band, tol in zip(golden_bands, unit_test_params.tols):
            # retrieve test and ref arrays for current band
            test_arr = test_ds.GetRasterBand(i_band).ReadAsArray()
            ref_arr = ref_ds.GetRasterBand(i_band).ReadAsArray()

            # calculate mean of absolute error and mask anything > 5.0
            err = np.abs(test_arr - ref_arr)
            err = np.ma.masked_array(err, mask=err > 5.0)
            mean_err = np.mean(err)

            # check if tolerances met
            assert mean_err < tol, f"band {i_band} of {test_path} mean err fail"

        del test_ds

File: python/extensions/pybind_isce3/cuda/geometry/rdr2geo.cpp
#include "rdr2geo.h"

#include <pybind11/numpy.h>
#include <string>

#include <isce3/core/Constants.h>
#include <isce3/core/Ellipsoid.h>
#include <isce3/core/LUT2d.h>
#include <isce3/core/Orbit.h>
#include <isce3/io/Raster.h>
#include <isce3/product/RadarGridParameters.h>

using isce3::cuda::geometry::Topo;

namespace py = pybind11;

void addbinding(py::class_<Topo> & pyRdr2Geo)
{
    pyRdr2Geo
            .def(py::init([](const isce3::product::RadarGridParameters&
                                          radar_grid,
                                  const isce3::core::Orbit& orbit,
                                  const isce3::core::Ellipsoid& ellipsoid,
                                  const isce3::core::LUT2d<double>& doppler,
                                  const double threshold, const int numiter,
                                  const int extraiter,
                                  const dataInterpMethod dem_interp_method,
                                  const int epsg_out, const bool compute_mask,
                                  const int lines_per_block) {
                auto rdr2geo_obj = Topo(radar_grid, orbit, ellipsoid, doppler);
                rdr2geo_obj.threshold(threshold);
                rdr2geo_obj.numiter(numiter);
                rdr2geo_obj.extraiter(extraiter);
                rdr2geo_obj.demMethod(dem_interp_method);
                rdr2geo_obj.epsgOut(epsg_out);
                rdr2geo_obj.computeMask(compute_mask);
                rdr2geo_obj.linesPerBlock(lines_per_block);
                return rdr2geo_obj;
            }),
                    py::arg("radar_grid"), py::arg("orbit"),
                    py::arg("ellipsoid"),
                    py::arg("doppler") = isce3::core::LUT2d<double>(),
                    py::arg("threshold") = 0.05, py::arg("numiter") = 25,
                    py::arg("extraiter") = 10,
                    py::arg("dem_interp_method") =
                            isce3::core::BIQUINTIC_METHOD,
                    py::arg("epsg_out") = 4326, py::arg("compute_mask") = true,
                    py::arg("lines_per_block") = 1000)
            .def("topo",
                    py::overload_cast<isce3::io::Raster&, const std::string&>(
                            &Topo::topo),
                    py::arg("dem_raster"), py::arg("outdir"))
            .def("topo",
                    py::overload_cast<isce3::io::Raster&,
                                      isce3::io::Raster*,
                                      isce3::io::Raster*,
                                      isce3::io::Raster*,
                                      isce3::io::Raster*,
                                      isce3::io::Raster*,
                                      isce3::io::Raster*,
                                      isce3::io::Raster*,
                                      isce3::io::Raster*,
                                      isce3::io::Raster*,
                                      isce3::io::Raster*,
                                      isce3::io::Raster*>
                            (&Topo::topo),
                    py::arg("dem_raster"),
                    py::arg("x_raster"),
                    py::arg("y_raster"),
                    py::arg("height_raster"),
                    py::arg("incidence_angle_raster"),
                    py::arg("heading_angle_raster"),
                    py::arg("local_incidence_angle_raster"),
                    py::arg("local_Psi_raster"),
                    py::arg("simulated_amplitude_raster"),
                    py::arg("layover_shadow_raster"),
                    py::arg("ground_to_sat_east_raster"),
                    py::arg("ground_to_sat_north_raster"),
                    R"(
        Run topo and output to user created topo rasters

        Parameters
        ----------
        dem_raster: isce3.io.Raster
            Input DEM raster
        x_raster: isce3.io.Raster
            Output raster for X coordinate in requested projection system (meters or degrees)
        y_raster: isce3.io.Raster
            Output raster for Y cooordinate in requested projection system (meters or degrees)
        height_raster: isce3.io.Raster
            Output raster for height above ellipsoid (meters)
        incidence_raster: isce3.io.Raster
            Output raster for incidence angle (degrees) computed from vertical at target
        heading_angle_raster: isce3.io.Raster
            Output raster for azimuth angle (degrees) computed anti-clockwise from EAST (Right hand rule)
        local_incidence_raster: isce3.io.Raster
            Output raster for local incidence angle (degrees) at target
        local_psi_raster: isce3.io.Raster
            Output raster for local projection angle (degrees) at target
        simulated_amplitude_raster: isce3.io.Raster
            Output raster for simulated amplitude image.
        layover_shadow_raster: isce3.io.Raster
            Output raster for layover/shadow mask.
        ground_to_sat_east_raster: isce3.io.Raster
            Output raster for east component of ground to satellite unit vector
        ground_to_sat_north_raster: isce3.io.Raster
            Output raster for north component of ground to satellite unit vector
                    )")
            .def_property_readonly("orbit", &Topo::orbit)
            .def_property_readonly("ellipsoid", &Topo::ellipsoid)
            .def_property_readonly("doppler", &Topo::doppler)
            .def_property_readonly("radar_grid", &Topo::radarGridParameters)
            .def_property("threshold",
                    py::overload_cast<>(&Topo::threshold, py::const_),
                    py::overload_cast<double>(&Topo::threshold))
            .def_property("numiter",
                    py::overload_cast<>(&Topo::numiter, py::const_),
                    py::overload_cast<int>(&Topo::numiter))
            .def_property("extraiter",
                    py::overload_cast<>(&Topo::extraiter, py::const_),
                    py::overload_cast<int>(&Topo::extraiter))
            .def_property("dem_interp_method",
                    py::overload_cast<>(&Topo::demMethod, py::const_),
                    py::overload_cast<dataInterpMethod>(&Topo::demMethod))
            .def_property("epsg_out",
                    py::overload_cast<>(&Topo::epsgOut, py::const_),
                    py::overload_cast<int>(&Topo::epsgOut))
            .def_property("compute_mask",
                    py::overload_cast<>(&Topo::computeMask, py::const_),
                    py::overload_cast<bool>(&Topo::computeMask))
            .def_property("lines_per_block",
                    py::overload_cast<>(&Topo::linesPerBlock, py::const_),
                    py::overload_cast<size_t>(&Topo::linesPerBlock))
            ;
}
